File: earthsciio/provider.py
"""The cadence-aware **Provider** (component (b)) — the sanctioned impure I/O
boundary for one ESS ``DataLoader`` node.

A :class:`Provider` is bound to one loader and created once per simulation. It
resolves a URL (per cadence anchor, for a time-varying source), fetches it
through the content-addressed cache (component (a), :mod:`earthsciio.cache`),
decodes it with the loader's :data:`~earthsciio.registry.format_registry` reader,
and returns RAW native-grid arrays (:class:`~earthsciio.native.NativeDataset`).
Variable-name remap / ``unit_conversion`` stay in ESS; regrid is ESD/C4's job
(Risk R3) — the Provider returns native arrays and nothing more.

It provides DATA, not a solver. The library EXPOSES the provider and its
:meth:`Provider.refresh_times`; the **user/solver** drives the discrete-cadence
update (e.g. a ``DifferentialEquations`` ``PresetTimeCallback`` /
``diffeqpy``/SciPy event that calls :meth:`Provider.refresh` at each anchor). No
solver is embedded here. The refresh is wired to fire **once per cadence
boundary** in the solver callback — NEVER per-RHS — exactly matching campfire C1
(``ess-06y``) terminal-event segmentation; downstream is C1 (the cadence
callback) + C4 (the regrid handoff), integration verified campfire-side.

CONST vs DISCRETE (the cadence contract):

* **CONST** — ``loader.temporal is None``: :meth:`materialize` reads the single
  file once; :meth:`refresh_times` is empty; :meth:`refresh` returns that same
  constant data.
* **DISCRETE** — ``loader.temporal`` set: :meth:`refresh` snaps a time to the
  loader's cadence **anchor** and returns the matching record's native arrays;
  :meth:`refresh_times` is the cadence schedule over the run window (the solver
  tstops). A multi-record file is sliced on its ``time_dim`` axis; a file-per-tick
  source resolves a new URL per file period.

This mirrors the peer Rust ``Provider`` (``rust/src/provider.rs``) and Julia
``Provider`` (``julia/src/provider.jl``); the decoded native arrays are equal
across all three tracks (conformance ``esio-9nb.9``).
"""

from __future__ import annotations

import datetime as _dt
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional, Sequence, Tuple, Union

import numpy as np

from .cache import Cache, CacheEntry
from .native import NativeDataset, NativeField
from .registry import Registry, format_registry

__all__ = ["LoaderTemporal", "DataLoader", "Provider", "Window"]

#: A run window ``(start, end)`` — half-open ``[start, end)``.
Window = Tuple[_dt.datetime, _dt.datetime]

#: A URL spec: a literal URL, a ``strftime`` template (contains ``%``), or a
#: callable ``anchor -> url`` (the most general, per-file-anchor resolver).
UrlSpec = Union[str, Callable[[_dt.datetime], str]]

_EPOCH = _dt.datetime(1970, 1, 1, tzinfo=_dt.timezone.utc)
_ZERO = _dt.timedelta(0)


def _snap_down(start: _dt.datetime, t: _dt.datetime, step: _dt.timedelta) -> _dt.datetime:
    """The largest aligned anchor ``start + k*step <= t`` (``k`` an integer).

    Timedelta floor-division floors toward -infinity, so this snaps ``t`` *down*
    to its cadence anchor even before ``start`` (caller validates the range).
    """
    return start + ((t - start) // step) * step


@dataclass(frozen=True)
class LoaderTemporal:
    """The temporal cadence of a DISCRETE loader (absent ⇒ CONST).

    ``frequency`` is the cadence step (e.g. 1 hour for ERA5) — it drives the
    refresh tstops and which record within a file is current. ``file_period`` is
    the span of one file (e.g. 1 day) — it drives URL resolution; a file holds
    ``file_period / frequency`` records along ``time_dim``. Anchors are aligned
    to ``start`` (the loader epoch). ``end`` (optional) is the exclusive end of
    available data and bounds :meth:`Provider.refresh_times` when no run window
    is given.
    """

    start: _dt.datetime
    frequency: _dt.timedelta
    file_period: _dt.timedelta
    end: Optional[_dt.datetime] = None
    time_dim: str = "time"

    def __post_init__(self) -> None:
        if self.frequency <= _ZERO:
            raise ValueError(f"frequency must be positive, got {self.frequency!r}")
        if self.file_period <= _ZERO:
            raise ValueError(f"file_period must be positive, got {self.file_period!r}")


@dataclass
class DataLoader:
    """The I/O-relevant projection of an ESS ``DataLoader`` the Provider needs.

    The full ESM contract (units, variable remap, grid family) lives upstream in
    ESS; this is only what resolves bytes and decodes them.

    Parameters
    ----------
    name:
        Loader name — provenance, recorded in the cache manifest.
    format:
        Format-registry key selecting the reader (e.g. ``"netcdf"``, ``"csv"``).
    url:
        A literal URL (CONST), a ``strftime`` template resolved at each file
        anchor (e.g. ``".../era5/%Y/%m/%Y%m%d.nc"``), or a callable
        ``anchor -> url``.
    variables:
        On-disk ``file_variable`` names to read; empty ⇒ all data variables
        (coordinates are always kept).
    temporal:
        The cadence (``None`` ⇒ CONST/static).
    mirrors:
        Failover URL candidates sharing the same cache identity.
    auth_realm:
        Auth realm to fetch under (resolved by the cache's auth registry).
    reader_kwargs:
        Extra keywords forwarded to the reader's ``read_native`` (e.g. the CSV
        reader's ``numeric_columns``).
    """

    name: str
    format: str
    url: UrlSpec
    variables: Sequence[str] = ()
    temporal: Optional[LoaderTemporal] = None
    mirrors: Sequence[str] = ()
    auth_realm: Optional[str] = None
    reader_kwargs: Dict[str, object] = field(default_factory=dict)

    @property
    def is_const(self) -> bool:
        """True if the loader is time-invariant (no :class:`LoaderTemporal`)."""
        return self.temporal is None

    def resolve_url(self, anchor: _dt.datetime) -> str:
        """Resolve the file URL for a cadence/file anchor."""
        u = self.url
        if callable(u):
            return u(anchor)
        if "%" in u:
            return anchor.strftime(u)
        return u


class Provider:
    """A loader-bound provider of native-grid arrays, refreshed at the loader's
    cadence. See the module docstring for the CONST/DISCRETE contract.

    Parameters
    ----------
    loader:
        The :class:`DataLoader` this provider serves.
    cache:
        The content-addressed :class:`~earthsciio.cache.Cache` (component (a)).
    window:
        Optional run window ``(start, end)`` bounding :meth:`refresh_times` /
        :meth:`prefetch` and priming :meth:`materialize` for a DISCRETE loader.
    formats:
        The format registry to resolve the reader from (defaults to the global
        :data:`~earthsciio.registry.format_registry`) — the seam that lets a new
        format plug in with no Provider change.

    Raises :class:`~earthsciio.errors.BackendNotRegistered` if no reader is
    registered for ``loader.format``.
    """

    def __init__(
        self,
        loader: DataLoader,
        cache: Cache,
        window: Optional[Window] = None,
        *,
        formats: Optional[Registry] = None,
    ) -> None:
        self.loader = loader
        self.cache = cache
        self.window = window
        registry = formats if formats is not None else format_registry
        # Resolve (and construct) the reader by name now, so an unknown format
        # fails at construction, not mid-solve.
        self._reader = registry.create(loader.format)
        # Current state — a file-read cache so stepping within one file decodes
        # it once, and the buffer the solver reads between cadence boundaries.
        self._current: Optional[NativeDataset] = None
        self._current_anchor: Optional[_dt.datetime] = None
        self._current_file_anchor: Optional[_dt.datetime] = None
        self._current_file: Optional[NativeDataset] = None

    # -- introspection ------------------------------------------------------ #

    @property
    def is_const(self) -> bool:
        """True if the bound loader is time-invariant (CONST)."""
        return self.loader.temporal is None

    @property
    def coords(self) -> Dict[str, NativeField]:
        """Native coordinates of the current buffer (lat/lon/…); empty until the
        first :meth:`materialize`/:meth:`refresh`."""
        return self._current.coords if self._current is not None else {}

    # -- materialize / refresh --------------------------------------------- #

    def materialize(self) -> NativeDataset:
        """Materialize the loader's native arrays into the buffer and return them.

        CONST: reads the single file once. DISCRETE: primes the buffer at the
        first cadence anchor of the window (≡ ``refresh(window.start)``), so a
        caller can read an initial state before stepping.
        """
        if self.loader.temporal is None:
            ds = self._read_file(self.loader.resolve_url(_EPOCH))
            self._current = ds
            return ds
        return self.refresh(self._lower_bound())

    def refresh(self, t: _dt.datetime) -> NativeDataset:
        """Refresh the buffer to the cadence anchor for time ``t`` and return it.

        Snaps ``t`` down to the loader's cadence anchor, re-reads the covering
        file only when the **file** anchor changed (stepping within one file
        decodes it once), slices the current record on ``time_dim``, and returns
        the native arrays. For a CONST loader this returns the constant data
        (materializing once if needed). Raises :class:`ValueError` if ``t``
        precedes the loader epoch, :class:`IndexError` if the snapped record is
        absent from its file.
        """
        temporal = self.loader.temporal
        if temporal is None:
            return self._current if self._current is not None else self.materialize()

        anchor = _snap_down(temporal.start, t, temporal.frequency)
        if anchor < temporal.start:
            raise ValueError(
                f"t={t!r} precedes the loader start {temporal.start!r}"
            )
        file_anchor = _snap_down(temporal.start, anchor, temporal.file_period)
        if self._current_file is None or self._current_file_anchor != file_anchor:
            self._current_file = self._read_file(self.loader.resolve_url(file_anchor))
            self._current_file_anchor = file_anchor

        rec = (anchor - file_anchor) // temporal.frequency
        sliced = _slice_record(self._current_file, temporal.time_dim, rec)
        self._current = sliced
        self._current_anchor = anchor
        return sliced

    def refresh_times(self) -> List[_dt.datetime]:
        """The cadence anchors at which the data changes and the solver must
        :meth:`refresh` — the solver tstops.

        Empty for a CONST loader, or for an unbounded DISCRETE loader with no run
        window and no ``temporal.end``. Otherwise the aligned anchors in
        ``[lower, upper)`` where ``lower`` is the window start clamped to the
        loader epoch and ``upper`` is the window end (else ``temporal.end``).
        Each is a timezone-aware :class:`datetime.datetime`; a solver integrating
        in epoch seconds takes ``t.timestamp()``.
        """
        temporal = self.loader.temporal
        if temporal is None:
            return []
        upper = self.window[1] if self.window is not None else temporal.end
        if upper is None:
            return []  # unbounded — no enumerable schedule
        lower = self._lower_bound()
        # First aligned anchor >= lower (ceil of the elapsed step count).
        steps = -((temporal.start - lower) // temporal.frequency)
        anchor = temporal.start + steps * temporal.frequency
        out: List[_dt.datetime] = []
        while anchor < upper:
            out.append(anchor)
            anchor = anchor + temporal.frequency
        return out

    # -- prefetch ----------------------------------------------------------- #

    def prefetch(self, window: Optional[Window] = None) -> List[CacheEntry]:
        """Warm the cache for every file the provider will need, WITHOUT decoding.

        CONST: the single file. DISCRETE: each unique file covering the window
        (``window`` arg, else the provider's window). Lets a caller pull all
        blobs up front (e.g. before a solve, or while online) so later
        :meth:`materialize`/:meth:`refresh` hit a warm, offline-readable cache.
        Returns the :class:`~earthsciio.cache.CacheEntry` for each unique file.
        """
        temporal = self.loader.temporal
        if temporal is None:
            return [self._fetch(self.loader.resolve_url(_EPOCH))]

        win = window if window is not None else self.window
        upper = win[1] if win is not None else temporal.end
        if upper is None:
            raise ValueError(
                "prefetch needs a bounded window: pass window=(start, end) or set "
                "LoaderTemporal.end"
            )
        start = win[0] if (win is not None and win[0] > temporal.start) else temporal.start
        entries: List[CacheEntry] = []
        seen: set = set()
        file_anchor = _snap_down(temporal.start, start, temporal.file_period)
        while file_anchor < upper:
            url = self.loader.resolve_url(file_anchor)
            if url not in seen:
                seen.add(url)
                entries.append(self._fetch(url))
            file_anchor = file_anchor + temporal.file_period
        return entries

    # -- internals ---------------------------------------------------------- #

    def _lower_bound(self) -> _dt.datetime:
        """The effective lower bound: the window start clamped to the loader
        epoch (a DISCRETE provider always has a temporal)."""
        temporal = self.loader.temporal
        assert temporal is not None
        if self.window is not None and self.window[0] > temporal.start:
            return self.window[0]
        return temporal.start

    def _fetch(self, url: str) -> CacheEntry:
        return self.cache.fetch(
            url,
            source_loader=self.loader.name,
            auth_realm=self.loader.auth_realm,
            mirrors=tuple(self.loader.mirrors),
        )

    def _read_file(self, url: str) -> NativeDataset:
        entry = self._fetch(url)
        handle = self._reader.open(entry.path)
        variables = list(self.loader.variables) if self.loader.variables else None
        return self._reader.read_native(handle, variables, **self.loader.reader_kwargs)


def _slice_record(ds: NativeDataset, time_dim: str, rec: int) -> NativeDataset:
    """Slice cadence record ``rec`` out of every field carrying ``time_dim``.

    Non-temporal fields pass through whole; the sliced ``time_dim`` coordinate is
    dropped (its value for this record is implied by the anchor). Mirrors the
    Julia ``_slice_dim`` / Rust ``select_leading``.
    """
    variables = {name: _slice_field(f, time_dim, rec) for name, f in ds.variables.items()}
    coords = {
        name: _slice_field(f, time_dim, rec)
        for name, f in ds.coords.items()
        if name != time_dim
    }
    return NativeDataset(variables, coords)


def _slice_field(f: NativeField, time_dim: str, rec: int) -> NativeField:
    if time_dim not in f.dims:
        return f  # non-temporal field unchanged
    axis = f.dims.index(time_dim)
    length = f.data.shape[axis]
    if rec < 0 or rec >= length:
        raise IndexError(
            f"cadence record {rec} out of range for dimension {time_dim!r} "
            f"(length {length})"
        )
    sliced = np.take(f.data, rec, axis=axis)
    new_dims = tuple(d for d in f.dims if d != time_dim)
    return NativeField(sliced, new_dims, f.attrs)

File: julia/src/provider.jl
# The cadence Provider — component (b) (esio-9nb.5). A Provider fulfils the ESS
# data-loader CONTRACT for one `.esm` node: it resolves a URL (per time, for a
# time-varying source), fetches it through the content-addressed cache
# (component a), decodes it with the named `FORMAT_REGISTRY` reader, and returns
# RAW native-grid arrays. Variable remap / unit conversion / regrid are NOT here
# — they stay in ESS/ESD (Risk R3).
#
# It provides DATA, not a solver (the sanctioned impure I/O boundary). The
# library EXPOSES the provider and its `refresh_times`; the USER/solver drives
# the discrete-cadence update (e.g. a DifferentialEquations `PresetTimeCallback`
# that calls `refresh(provider, t)` at each tick). No solver is embedded here —
# `library-exposes-rhs-not-solver`.

"""
    Cadence

Temporal cadence of a provider's data (the `.esm` node's `temporal-cadence`):

  * `CONST` — time-invariant. `refresh_times` is empty; the data is materialized
    once and never refreshed.
  * `DISCRETE` — changes at known time points. `refresh_times` returns that grid;
    the solver refreshes at each tick."""
@enum Cadence CONST DISCRETE

"""
    Provider(cache, url; format, cadence=CONST, times=Float64[], time_dim=nothing,
             variables=nothing, reader_kwargs=NamedTuple(),
             source_loader=nothing, auth_realm=nothing)

A data provider over the shared cache.

  * `url` — a resolved-URL `String` (constant source) OR a function `t -> url`
    (a time-varying source whose file/key changes per tick).
  * `format` — reader name in [`FORMAT_REGISTRY`] (`"netcdf"`, `"csv"`, …).
  * `cadence` — [`CONST`] or [`DISCRETE`]. `DISCRETE` requires non-empty `times`;
    `CONST` requires empty `times`.
  * `times` — the discrete cadence grid (sorted on construction); what
    [`refresh_times`] returns.
  * `time_dim` — when set on a `DISCRETE` provider whose single file holds the
    whole cadence on an internal axis (e.g. a daily file of hourly steps),
    [`materialize`]`(p, t)` slices that dimension to the tick's record.
  * `records_per_sample` — `nothing` or `1` (default) returns the SINGLE
    at-or-before record with `time_dim` DROPPED (held piecewise-constant); `2`
    returns the TWO bracketing records (floor + successor) with `time_dim`
    RETAINED at length 2 and a canonical 2-element `time_dim` coordinate of Unix
    epoch seconds, so a downstream model interpolates in time. `2` requires a
    `time_dim`; the successor is read across a file boundary when needed, and at
    the last cadence tick the bracket degenerates to `[last, last]` (equal
    timestamps) so the downstream weight clamps. The provider does pure I/O
    (returns N records) and performs no interpolation itself.
  * `variables` — restrict the returned data variables (coords always kept);
    `nothing` returns all.
  * `reader_kwargs` — extra keywords forwarded to [`read_native`] (e.g. the CSV
    reader's `numeric_columns`).
  * `source_loader` / `auth_realm` — recorded in the cache manifest on fetch."""
struct Provider
    cache::Cache
    format::String
    cadence::Cadence
    times::Vector{Float64}
    url_for::Function
    time_dim::Union{Nothing,String}
    variables::Union{Nothing,Vector{String}}
    reader_kwargs::Dict{Symbol,Any}
    source_loader::Union{Nothing,String}
    auth_realm::Union{Nothing,String}
    records_per_sample::Union{Nothing,Int}
end

function Provider(cache::Cache, url; format::AbstractString,
                  cadence::Cadence = CONST, times = Float64[],
                  time_dim = nothing, variables = nothing,
                  reader_kwargs = NamedTuple(),
                  source_loader = nothing, auth_realm = nothing,
                  records_per_sample = nothing)
    haskey(FORMAT_REGISTRY, format) || throw(ArgumentError(
        "format '$format' is not registered in the format registry; " *
        "registered: $(registered_names(FORMAT_REGISTRY))"))
    tvec = sort!(Float64[t for t in times])
    if cadence == CONST && !isempty(tvec)
        throw(ArgumentError(
            "a CONST provider has no refresh times, got $(length(tvec)) — use cadence=DISCRETE"))
    elseif cadence == DISCRETE && isempty(tvec)
        throw(ArgumentError(
            "a DISCRETE provider requires a non-empty cadence (times=...)"))
    end
    time_dim === nothing || cadence == DISCRETE ||
        throw(ArgumentError("time_dim only applies to a DISCRETE provider"))
    records_per_sample === nothing || records_per_sample == 1 || records_per_sample == 2 ||
        throw(ArgumentError(
            "records_per_sample must be 1 or 2, got $(repr(records_per_sample))"))
    records_per_sample != 2 || time_dim !== nothing ||
        throw(ArgumentError(
            "records_per_sample=2 needs a time_dim to bracket along"))
    url_for = url isa AbstractString ? (let u = String(url); _ -> u; end) : url
    return Provider(cache, String(format), cadence, tvec, url_for,
                    time_dim === nothing ? nothing : String(time_dim),
                    variables === nothing ? nothing : String.(collect(variables)),
                    Dict{Symbol,Any}(pairs(reader_kwargs)),
                    source_loader === nothing ? nothing : String(source_loader),
                    auth_realm === nothing ? nothing : String(auth_realm),
                    records_per_sample === nothing ? nothing : Int(records_per_sample))
end

"""Thin constructor for a time-invariant ([`CONST`]) provider."""
const_provider(cache::Cache, url; kwargs...) =
    Provider(cache, url; cadence = CONST, kwargs...)

"""Thin constructor for a time-varying ([`DISCRETE`]) provider over `times`."""
discrete_provider(cache::Cache, url, times; kwargs...) =
    Provider(cache, url; cadence = DISCRETE, times = times, kwargs...)

"""
    refresh_times(p::Provider) -> Vector{Float64}

The discrete time points at which the data changes and the solver must
[`refresh`]. Empty for a [`CONST`] provider; the sorted cadence grid for a
[`DISCRETE`] one. The library exposes these; the user wires them into the
solver (e.g. `PresetTimeCallback(refresh_times(p), …)`)."""
refresh_times(p::Provider) = copy(p.times)

"""True if `p`'s data is time-invariant ([`CONST`])."""
is_const(p::Provider) = p.cadence == CONST

# Fetch (cache) + decode (format reader) for the URL resolved at `t`.
function _load(p::Provider, t)
    entry = fetch_blob(p.cache, p.url_for(t);
                       source_loader = p.source_loader, auth_realm = p.auth_realm)
    nds = read_native(FORMAT_REGISTRY[p.format], entry.path; p.reader_kwargs...)
    return p.variables === nothing ? nds : _select(nds, p.variables)
end

function _select(nds::NativeDataset, want::Vector{String})
    keep = Dict{String,NativeField}()
    for name in want
        haskey(nds.variables, name) || throw(ArgumentError(
            "requested variable '$name' not in blob; present: $(variable_names(nds))"))
        keep[name] = nds.variables[name]
    end
    return NativeDataset(keep, nds.coords)
end

# Index of the cadence tick active at `t`: exact match, else the last tick ≤ t
# (the currently-in-effect record). Errors if `t` precedes the first tick.
function _tick_index(p::Provider, t::Real)
    tf = Float64(t)
    i = findfirst(==(tf), p.times)
    i === nothing || return i
    j = searchsortedlast(p.times, tf)
    j >= 1 && return j
    throw(ArgumentError(
        "t=$t precedes the provider's first refresh time $(first(p.times))"))
end

# Slice `dim` out of every variable that carries it, at record `idx`; drop the
# now-singular dimension and its coordinate. Used for an internal-axis DISCRETE
# source (one file, many time records).
function _slice_dim(nds::NativeDataset, dim::String, idx::Integer)
    vars = Dict{String,NativeField}()
    for (name, f) in nds.variables
        pos = findfirst(==(dim), f.dims)
        if pos === nothing
            vars[name] = f
        else
            sliced = collect(selectdim(f.data, pos, idx))
            vars[name] = NativeField(sliced, [d for d in f.dims if d != dim], f.attrs)
        end
    end
    coords = Dict(k => v for (k, v) in nds.coords if k != dim)
    return NativeDataset(vars, coords)
end

# --- 2-record bracket mode (records_per_sample=2) ---------------------------
# Mirrors the Python `Provider._refresh_bracket`: return the TWO records that
# bracket `t` (the floor tick + its successor) with `time_dim` RETAINED at length
# 2 and a canonical 2-element epoch-seconds `time_dim` coordinate, so a downstream
# model interpolates in time. The floor record is `_tick_index` (the same data
# index `_slice_dim` uses); the successor is the next data index along the axis,
# or — when that overruns the file — record 1 of the next file (`url_for(t_next)`,
# re-decoded since the Julia provider keeps no file buffer). At the last cadence
# tick there is no successor: the bracket degenerates to `[last, last]` (equal
# timestamps) so the downstream weight clamps — bracket mode never throws at the
# end of data.

const _CF_UNIT_SECONDS = Dict{String,Float64}(
    "second" => 1.0, "seconds" => 1.0, "sec" => 1.0, "secs" => 1.0, "s" => 1.0,
    "minute" => 60.0, "minutes" => 60.0, "min" => 60.0, "mins" => 60.0,
    "hour" => 3600.0, "hours" => 3600.0, "hr" => 3600.0, "hrs" => 3600.0, "h" => 3600.0,
    "day" => 86400.0, "days" => 86400.0, "d" => 86400.0)

# Parse a CF reference date (the `<ref>` in "<unit> since <ref>") → Unix epoch
# seconds, or `nothing`. Handles "yyyy-mm-dd[ HH[:MM[:SS]]]" with an optional
# `T`/space separator and a trailing `Z`/`UTC`/±offset (treated as UTC).
function _parse_cf_reference(s)
    str = strip(replace(String(s), 'T' => ' '))
    str = strip(replace(str, r"\s*(Z|UTC|[+-]\d{2}:?\d{2}(:\d{2})?)\s*$" => ""))
    for fmt in (dateformat"yyyy-mm-dd HH:MM:SS", dateformat"yyyy-mm-dd HH:MM",
                dateformat"yyyy-mm-dd HH", dateformat"yyyy-mm-dd")
        dt = tryparse(DateTime, str, fmt)
        dt === nothing || return datetime2unix(dt)
    end
    return nothing
end

# Parse a CF "<unit> since <reference>" units string → (ref_epoch_seconds,
# unit_seconds), or `nothing` when it can't be decoded (non-"since" units, an
# unknown step, or an unparseable reference) — the caller then emits raw times.
function _cf_time_scale(units)
    units === nothing && return nothing
    m = match(r"^\s*([A-Za-z]+)\s+since\s+(.+?)\s*$", String(units))
    m === nothing && return nothing
    step = get(_CF_UNIT_SECONDS, lowercase(m.captures[1]), nothing)
    step === nothing && return nothing
    ref = _parse_cf_reference(m.captures[2])
    ref === nothing && return nothing
    return (ref, step)
end

# Convert a raw cadence-grid value to Unix epoch seconds via a decoded CF scale;
# with no scale (units absent/undecodable) fall back to the raw value (documented
# deviation from the epoch-seconds contract).
_raw_to_epoch(raw, scale) =
    scale === nothing ? Float64(raw) : (scale[1] + Float64(raw) * scale[2])

# The CF `units` string of the `dim` coordinate in `nds`, or `nothing`.
function _time_units(nds::NativeDataset, dim::String)
    haskey(nds.coords, dim) || return nothing
    return get(nds.coords[dim].attrs, "units", nothing)
end

# Length of `nds` along `dim` (from the coordinate, else the first variable
# carrying it); 0 if nothing carries it.
function _time_len(nds::NativeDataset, dim::String)
    if haskey(nds.coords, dim)
        c = nds.coords[dim]
        pos = findfirst(==(dim), c.dims)
        pos === nothing || return size(c.data, pos)
    end
    for f in values(nds.variables)
        pos = findfirst(==(dim), f.dims)
        pos === nothing || return size(f.data, pos)
    end
    return 0
end

# Stack record `i0` of `a0` and record `i1` of `a1` along axis `pos`, keeping that
# axis at length 2 (floor, then successor). Range-index each so the axis survives,
# then `cat` — this is uniform across same-file, cross-file, and degenerate cases.
_stack_bracket(a0, i0::Integer, a1, i1::Integer, pos::Integer) =
    collect(cat(selectdim(a0, pos, i0:i0), selectdim(a1, pos, i1:i1); dims = pos))

# Assemble the 2-record bracket: every variable carrying `dim` gets records
# (`nds0`,`i0`) and (`nds1`,`i1`) stacked to a size-2 `dim` axis (dims/order/attrs
# preserved); non-temporal variables and non-`dim` coords pass through from
# `nds0`; `dim` becomes a 2-element epoch-seconds coordinate `[t0, t1]`.
function _bracket_build(nds0::NativeDataset, i0::Integer,
                        nds1::NativeDataset, i1::Integer,
                        dim::String, t0::Float64, t1::Float64)
    vars = Dict{String,NativeField}()
    for (name, f0) in nds0.variables
        pos = findfirst(==(dim), f0.dims)
        if pos === nothing
            vars[name] = f0                                  # non-temporal: unchanged
        else
            stacked = _stack_bracket(f0.data, i0, nds1.variables[name].data, i1, pos)
            vars[name] = NativeField(stacked, copy(f0.dims), f0.attrs)
        end
    end
    coords = Dict{String,NativeField}()
    for (k, v) in nds0.coords
        k == dim && continue                                 # replaced below
        coords[k] = v
    end
    coords[dim] = NativeField(Float64[t0, t1], [dim],
        Dict{String,Any}("units" => "seconds since 1970-01-01T00:00:00Z",
                         "calendar" => "standard"))
    return NativeDataset(vars, coords)
end

function _bracket(p::Provider, t::Real)
    dim = p.time_dim::String
    rec0 = _tick_index(p, t)                 # floor tick == floor data index
    nds0 = _load(p, t)
    scale = _cf_time_scale(_time_units(nds0, dim))
    L = _time_len(nds0, dim)
    t0 = _raw_to_epoch(p.times[rec0], scale)

    if rec0 >= length(p.times)               # last tick / past end → degenerate
        i0 = min(rec0, L)                     # clamp to a valid record (never throw)
        return _bracket_build(nds0, i0, nds0, i0, dim, t0, t0)
    end

    t1 = _raw_to_epoch(p.times[rec0 + 1], scale)
    if rec0 + 1 <= L                         # successor in the same file
        return _bracket_build(nds0, rec0, nds0, rec0 + 1, dim, t0, t1)
    end
    nds1 = _load(p, p.times[rec0 + 1])       # successor is record 1 of the next file
    return _bracket_build(nds0, min(rec0, L), nds1, 1, dim, t0, t1)
end

"""
    materialize(p::Provider, t::Real) -> NativeDataset
    materialize(p::Provider) -> NativeDataset

Return the native arrays for the source at time `t`. For a [`DISCRETE`] provider
with `time_dim`, the internal cadence axis is sliced to `t`'s record — unless
`records_per_sample=2`, in which case the two bracketing records (floor +
successor) are returned with `time_dim` retained at length 2 and a 2-element
epoch-seconds `time_dim` coordinate (see the [`Provider`] `records_per_sample`
field). The no-argument form is for a [`CONST`] provider (a `DISCRETE` provider
must be given a time)."""
function materialize(p::Provider, t::Real)
    if p.records_per_sample == 2 && p.time_dim !== nothing
        return _bracket(p, t)
    end
    nds = _load(p, t)
    if p.time_dim !== nothing
        return _slice_dim(nds, p.time_dim, _tick_index(p, t))
    end
    return nds
end

function materialize(p::Provider)
    p.cadence == CONST || throw(ArgumentError(
        "a DISCRETE provider needs a time: materialize(p, t) / refresh(p, t)"))
    return materialize(p, 0.0)
end

"""
    refresh(p::Provider, t::Real) -> NativeDataset

Re-materialize at the cadence tick `t`. This is the call the solver makes from
its `PresetTimeCallback` at each of [`refresh_times`]`(p)`. Identical to
[`materialize`]`(p, t)`; named for the solver-side update site."""
refresh(p::Provider, t::Real) = materialize(p, t)

"""
    prefetch(p::Provider) -> Vector{CacheEntry}

Warm the cache for every URL the provider will need — the single URL for a
[`CONST`] provider, the unique per-tick URLs for a [`DISCRETE`] one — WITHOUT
decoding. Lets a caller pull all blobs up front (e.g. before a solve, or while
online) so later [`materialize`]/[`refresh`] calls hit a warm, offline-readable
cache. Returns the [`CacheEntry`] for each."""
function prefetch(p::Provider)
    urls = p.cadence == CONST ? String[p.url_for(0.0)] :
           unique(String[p.url_for(t) for t in p.times])
    return CacheEntry[fetch_blob(p.cache, u; source_loader = p.source_loader,
                                 auth_realm = p.auth_realm) for u in urls]
end
